cdap-app-templates/cdap-etl/cdap-data-pipeline/src/main/java/io/cdap/cdap/datapipeline/StageOperationsValidator.java

```diff
@@ -151,26 +152,55 @@ void validate() {
     // Here outout d from OP1 is redundant, since the d in output schema will always come from OP2.
     // However d will not be in the redundantOutputs map, as we only put the redundant fields if they
     // appear in input of some operation. Such redundancy should cause validation checks to fail.
-    Iterator<Map.Entry<String, List<String>>> iterator = unusedOutputs.entrySet().iterator();
+    Iterator<Map.Entry<String, List<FieldOperation>>> iterator = unusedOutputs.entrySet().iterator();
     while (iterator.hasNext()) {
-      Map.Entry<String, List<String>> next = iterator.next();
+      Map.Entry<String, List<FieldOperation>> next = iterator.next();
       String field = next.getKey();
-      List<String> origins = next.getValue();
+      List<FieldOperation> origins = next.getValue();
 
       if (stageOutputs.contains(field)) {
-        // iterator.remove();
         if (origins.size() > 1) {
           // size will be greater than 1 only if its redundant
           origins.remove(origins.size() - 1);
-          // unusedOutputs.put(field, origins.subList(0, origins.size() - 1));
         } else {
           iterator.remove();
         }
+      } else {
+        // If this field is not in the output schema of the stage, it can still be valid if the operations which
+        // generate it are all input schema fields. For example, a Joiner joins two datasets D1,D2 based on the
+        // joiner key D1.K1, D2.K2, and decides to drop the joiner key in the output schema. The operation
+        // [D1.K1, D2.K2] ->[K1, K2] is a valid even though K1,K2 are not in the output schema. In this case, we need
+        // to check if the input fields of the operation are all from the input schema of the stage.
+        if (origins.size() != 1) {
+          continue;
```

Comment:
why does it have to be 1?

Reply:
Isn't it actually 2 in the example of a join?

Reply:
No in joiner it is also 1, the origin is corresponds to operations, not fields. If there are two operations that generate the field, it means it is redundant. If it is 0, it means this is a generated field, which we will deal with in the future prs.

Reply:
Oh I see, the origin is an operation. So as an example, if there are two operations that generate the same field - say, a projection that first renames A to X and then also renames B to X, what should be the behavior? Seems like the second operation prevails? 

In wrangler you could even have this scenario: 1) parse-as-csv creates "name" and other fields, 2) split "name" into "first" and "last", 3) drop "name" 4) rename "last" into "name". Then we would have 2 operations that output "name". What would be the correct behavior then?

Reply:
For the first case, the second operation always prevails, but we currently do not allow this kind of use case. We will consider the first operation redundant, so the validation should fail.
For second case, I don't know what should the correct field operation for drop a field. But basically the wrangler will emit operations like:
[body] -> [name, other fields],
[name] -> [first, last]
[last] -> [name],
this is valid in the current validation logic, since name in OP2 is used as input, so the OP3 will be only operation that outputs name, which is valid.

Reply:
Created the JIRA as we discussed https://issues.cask.co/browse/CDAP-15785